src/sqlfluff/core/templaters/jinja.py

```diff
@@ -829,26 +837,53 @@ def _handle_unreached_code(
                         if idx in uncovered_slices
                     )
 
-                    variants[variant_raw_str] = (score, trace)
+                    variants[variant_raw_str] = (score, trace, length_deltas)
 
         # Return the top-scoring variants.
-        sorted_variants: List[Tuple[int, JinjaTrace]] = sorted(
+        sorted_variants: List[Tuple[int, JinjaTrace, Dict[int, int]]] = sorted(
             variants.values(), key=lambda v: v[0], reverse=True
         )
-        for _, trace in sorted_variants[:max_variants_returned]:
-            # :TRICKY: Yield variants that _look like_ they were rendered from
-            # the original template, but actually were rendered from a modified
-            # template. This should ensure that lint issues and fixes for the
-            # variants are handled correctly and can be combined with those from
-            # the original template.
-            # To do this we run through modified slices and adjust their source
-            # slices to correspond with the original version. We do this by referencing
-            # their slice position in the original file, because we know we haven't
-            # changed the number or ordering of slices, just their length/content.
-            adjusted_slices: List[TemplatedFileSlice] = [
-                tfs._replace(source_slice=original_source_slices[idx])
-                for idx, tfs in enumerate(trace.sliced_file)
-            ]
+        for _, trace, deltas in sorted_variants[:max_variants_returned]:
+            # :TRICKY: We want to yield variants that _look like_ they were
+            # rendered from the original template. However, they were actually
+            # rendered from a modified template, which means they have source
+            # indices which won't line up with the source files. We correct that
+            # here by using the length deltas generated earlier from the
+            # modifications.
+            # This should ensure that lint issues and fixes for the variants are
+            # handled correctly and can be combined with those from the original
+            # template.
+            delta_stack = sorted(deltas.items(), key=lambda t: t[0])
```

Comment:
Could you add a comment explaining the purpose of the sorting, and what you're sorting on?

Given the amount of code being added here, I suggest moving this bit to a new function for readability.

Reply:
Good idea